iree/compiler/Dialect/LinalgExt/Transforms/Tiling.cpp

```diff
@@ -0,0 +1,415 @@
+    if (tiledOp->getNumResults()) {
+      SmallVector<Value> results;
+      results.reserve(tiledOp->getNumResults());
+      for (auto en : llvm::enumerate(tiledOp->getResults())) {
+        Value result = en.value();
+        ArrayRef<OpFoldResult> offsets(resultOffsets[en.index()]);
+        auto resultType = result.getType().cast<ShapedType>();
+        auto oneAttr = builder.getI64IntegerAttr(1);
+        SmallVector<OpFoldResult> strides(resultType.getRank(), oneAttr);
+        auto sizes = llvm::to_vector<4>(llvm::map_range(
+            llvm::seq<int64_t>(0, resultType.getRank()),
+            [&](int64_t dim) { return getDim(builder, loc, result, dim); }));
+        Value insert = builder.create<tensor::InsertSliceOp>(
+            loc, result, outputs[en.index()], offsets, sizes, strides);
+        results.push_back(insert);
+      }
+      std::swap(ret.results, results);
+    }
+    return ret;
+  }
+
+  // If tile size at this depth is empty, do nothing.
+  if (isUntiledLoop(tileSizes[loopDepth])) {
+    auto zeroAttr = builder.getI64IntegerAttr(0);
+    offsets.push_back(zeroAttr);
+    assert(matchPattern(loopBounds[loopDepth].offset, m_Zero()) &&
+           "expected loop bounds to have lower bound of zero");
+    tileSizes[loopDepth] = getOpFoldResult(loopBounds[loopDepth].size);
+    return tileLinalgExtOpImpl(builder, op, outputs, tileSizes, iteratorTypes,
+                               loopBounds, loopDepth + 1, offsets,
+                               distributionInfo);
+  }
+
+  // Generate an scf.for for the current loop depth.
+  Value lb = loopBounds[loopDepth].offset;
+  Value ub = loopBounds[loopDepth].size;
+  if (!matchPattern(loopBounds[loopDepth].stride, m_One())) {
+    return static_cast<LogicalResult>(
+        op.emitOpError("expected stride to be 1"));
+  }
+  Value step = getValue(builder, loc, tileSizes[loopDepth]);
+
+  // Update lb, ub and step for cyclic distribution.
+  if (!distributionInfo.empty() &&
+      iteratorTypes[loopDepth] == getParallelIteratorTypeName()) {
+    linalg::updateBoundsForCyclicDistribution(
+        builder, loc, distributionInfo.front().procId,
+        distributionInfo.front().nprocs, lb, ub, step);
+    distributionInfo = distributionInfo.drop_front();
+  }
+  FailureOr<TiledOp> innerReturnValue;
+  bool isBufferTiling = op->getNumResults() == 0;
+  ValueRange initValues(isBufferTiling ? ValueRange{} : outputs);
+  auto forOp = builder.create<scf::ForOp>(
+      loc, lb, ub, step, initValues,
+      [&](OpBuilder &b, Location loc, Value iv, ValueRange args) {
+        offsets.push_back(iv);
+        auto affineMaps = AffineMap::inferFromExprList({ArrayRef<AffineExpr>{
+            b.getAffineSymbolExpr(0),
+            b.getAffineSymbolExpr(1) - b.getAffineDimExpr(0)}})[0];
+        Value inBoundsTileSize = b.create<AffineMinOp>(
+            loc, affineMaps,
+            ValueRange{iv, getValue(builder, loc, tileSizes[loopDepth]), ub});
```

Comment:
[optional] It took me a bit time to parse the check. Maybe we can add a comment like `The in bounds tile size is min(step, ub- iv).`?

Also, I know they are the same, but I'd prefer using `step` instead of `getValue(builder, loc, tileSizes[loopDepth])`. It is clear to me that step is tile size, and we don't have to build the Value if it is an attribute.

Reply:
The `step` might not be the same as `tileSizes[loopDepth]` since the distribution might make the step `tileSizes[loopDepth] * nprocs`.

Reply:
I see, thanks!